### ISSUE TITLE
Port twisted.web.client.HTTPClientFactory to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6077]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6077|
|Type|enhancement|
|Created|2012-10-08 11:03:04Z|
|Branch|https://github.com/twisted/twisted/tree/webclient-py3-6077-2|

This is the basis of the old web client.

<details><summary>Searchable metadata</summary>

```
trac-id__6077 6077
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_webclient_py3_6077_2 branches/webclient-py3-6077-2
branch_author__itamarst__exarkun itamarst, exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1349694184000000 1349694184000000
changetime__1354219070000000 1354219070000000
version__None None
owner__itamar itamar
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [35999]) Branching to 'webclient-py3-6077'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

(In [36411]) Branching to 'webclient-py3-6077'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [36412]) Move Agent tests into new test module. Refs [#6077](https://github.com/twisted/twisted/issues/6077)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

(In [36434]) Branching to 'webclient-py3-6077-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set owner to @exarkun|
|-|-|

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/webclient-py3-6077-2

The code in the new `test_agent.py` is identical to the tests for Agent that were previously in `test_webclient.py`, so not much to review there.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. Doc for `code` parameter to `Error.__init__` has a typo - _bytesing_
  1. It seems type check for `code` in `Error.__init__` is untested.  I deleted it and all the tests still pass.  Also I didn't like the look of it, so hopefully it _can_ just be deleted?
  1. in `client.py`, there seems to be a mix of two approaches to formatting bytes:
    * `bytes + b'x' + bytes`
    * `networkString('%sx%s' % (nativeString(bytes), nativeString(bytes))`
  This isn't really a problem I guess, but it might be nice to decide on one approach and stick to it consistently.  If there's some particular reason to prefer one or the other of these in different circumstances, please disregard (but maybe document it on the porting page?)
  1. I guess it would be good to have a ticket for dealing with the `downloadPage` test skips somehow.  Delete `downloadPage`?  I forget whether `Agent` is feature equivalent yet.  Or... port `downloadPage`.  And referenced from the code in the typical manner.
  1. Blech on `testRawSomeCookies` :(

That's it.  If you can address all these (save the very last point) with nothing too surprising, please merge.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

1.Fixed.
 2. I think that means there's a missing test. It's not exactly important code though, so I'll just chalk it up to the large list of missing test coverage for twisted.web. Deleted.
 3. I switched to first format. My gut feeling is "use networkString variant if it's sufficiently complex that it's clearer than lots of addition", which these cases definitely aren't.
 4. [#6197](https://github.com/twisted/twisted/issues/6197), added reference.

Left the icky test in, which I think you implied. Plus side, I think we're maybe one feature away (browser-style redirects) from being able to rewrite getPage and downloadPage as wrappers on top of Agent.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

(In [36468]) Address review comments. Refs [#6077](https://github.com/twisted/twisted/issues/6077)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set status to `closed`|
|-|-|

(In [36470]) Merge webclient-py3-6077-2.

Fixes: [#6077](https://github.com/twisted/twisted/issues/6077)
Author: itamar
Review: exarkun

Port twisted.web.client.getPage to Python 3. Along the way, moved Agent tests into their own test module.
